README
Sys::Hwloc
==========

This is a perl binding for the Portable Hardware Locality (hwloc) library.
See http://www.open-mpi.org/projects/hwloc/ for details.


DEPENDENCIES
------------

This module requires hwloc >= 0.9 (recommended: >= 1.1)

Makefile.PL uses ExtUtils::MakeMaker and ExtUtils::Constant
to create the Makefile.

Parts of the test suite need Test::More >= 0.89.

The module was tested on some i686/intel64 machines
running recent Linux and Perl 5.8.8 or 5.10.

Other architectures were not tested.

The Windows OS is currently not supported.


INSTALLATION
------------

To install this module, type the following:

   perl Makefile.PL
   make
   make test
   make install

hwloc libraries and header files are searched for by using pkg-config.
If they cannot be found, set the PKG_CONFIG_PATH environment variable
to the directory where the file hwloc.pc is located.


EXAMPLE SCRIPTS
---------------

Look into the examples directory to get inspired.


HWLOC API VERSION
-----------------

The configure step figures out the value of the HWLOC_API_VERSION constant
from hwloc.h, and compiles and exports Hwloc functions accordingly.
Thus, the Hwloc module depends on the hwloc version it was built against.
Undefined results are got when another hwloc library version is used at
runtime.

This constant was introduced in hwloc 1.0. For earlier hwloc versions,
the Hwloc module exports this constant with an undefined value. For convenience,
the Hwloc module also exports the constant HWLOC_XSAPI_VERSION, which
is always defined.

These constants can be used in a Perl script using Hwloc as follows to make it
work with different HWLOC_API_VERSIONs:

$proc_t = HWLOC_API_VERSION() ? HWLOC_OBJ_PU() : HWLOC_OBJ_PROC();

To bind a Perl script to a specific hwloc API version, check it
in a BEGIN block:

BEGIN {
  die "This script needs HWLOC_XSAPI_VERSION >= 0x00010000\n" unless
    HWLOC_XSAPI_VERSION() >= 0x00010000;
}

The compatibility matrix of released hwloc versions with Sys::Hwloc is

   0.9.1   not tested
   0.9.2   works (some tests fail due to unspecified C API behaviour)
   0.9.3   works (some tests fail due to unspecified C API behaviour)
   1.0     not tested
   1.0.1   not tested
   1.0.2   works
   1.0.3   works
   1.1     works (tests 08-bitmap 2,59 fail due to hwloc bug)



XML SUPPORT
-----------

The configure step tries to figure out if the hwloc library was
built with XML support.

Depending on the result, Hwloc functions dealing with XML
are compiled and exported or not, and the special constant
Sys::Hwloc::HWLOC_HAS_XML is exported with a true or false value.
It can be used in Perl scripts using Hwloc as follows:

if(HWLOC_HAS_XML) {
  $topology->export_xml('-');
} else {
  die "Your hwloc lib was built without XML support";
}


CPUSETS, NODESETS, BITMAPS
--------------------------

In hwloc 0.9 and 1.0 a hwloc_obj struct defines
the struct members cpuset and nodeset with C type hwloc_cpuset.
Data of type hwloc_cpuset become created and manipulated with
functions of the Cpuset API.
When build with these hwloc versions, the Sys::Hwloc module exports
the functions of the Cpuset API, blesses these data into the
namespace Sys::Hwloc::Cpuset, and provides OO-ish methods for them.

In hwloc 1.1 a hwloc_obj struct defines struct members with
C type hwloc_cpuset, and struct members with C type hwloc_nodeset.
Both C types are aliases of the C type hwloc_bitmap. These
data become created and manipulated with functions of the Bitmap API.
When built with these hwloc versions, the Sys::Hwloc module
exports the functions of the Bitmap API, blesses these data
into the namespace Sys::Hwloc::Bitmap, and provides OO-ish
methods for them. A distrinction between the C types hwloc_cpuset
and hwloc_nodeset is not made.


TESTING
-------

The test suite runs tests against the hwloc topology of the current
machine.

Arbitrary topologies can be tested when Hwloc was compiled
with XML support. Set the HWLOC_XMLFILE environment variable
to the path of an XML file describing the topology,
and run the test suite.

The test suite makes assumptions about what C API functions should
return, based on the docs and observed behaviour of hwloc-1.0.

Some tests may apparently fail when testing return values,
if other versions are used and the hwloc C API behaves differently.


TODO
----

- extend XML support by using XML::LibXML


BUGS
----

Note that this is a wrapper, which cannot correct bugs in the underlying
hwloc library.

If you feel that you found a wrapper bug, report it via
https://rt.cpan.org/

If you feel that you found a hwloc bug, refer to
http://www.open-mpi.org/projects/hwloc/ how to report it.


COPYRIGHT AND LICENSE
---------------------

Copyright (C) 2010 Zuse Institute Berlin

This package and its accompanying libraries is free software; you can
redistribute it and/or modify it under the terms of the GPL version 2.0,
or the Artistic License 2.0. Refer to LICENSE for the full license text.


Changes
Revision history for Perl extension Sys::Hwloc.

0.08 2011/01/05
      - bug fix prototype hwloc_get_next_child, added corresponding test
      - prepared for upcoming availability of hwloc_bitmap_list_snprintf/sscanf
        by renaming
          hwloc_bitmap_sprintf_list -> hwloc_bitmap_list_sprintf
          hwloc_bitmap_sscanf_list  -> hwloc_bitmap_list_sscanf
        (internally still using our own implementation)
      - added module "Finding Objects Inside a CPU set"

0.07  2010/12/29
      - implemented CPU binding
      - implemented Memory binding
      - added support for upcoming hwloc_get_api_version() check
      - refined Makefile.PL (use %Config for preproc and compile/link)
      - not export Cpuset API, Bitmap API, Binding API by default,
        but by %EXPORT_TAGS going into @EXPORT_OK

0.06  2010/12/22
      - corrected typos in PODs
      - added dependency on ExtUtils::Constant in Makefile.PL
      - added high-level cpuset/nodeset tests on current topology
      - added examples/gen_cpumapping (useful to set affinity maps
        for MPI apps)

0.05  2010/12/21
      - added Sys::Hwloc::Obj->sprintf() as additional alias for
        hwloc_obj_sprintf() and Sys::Hwloc::Topology->sprintf_obj()

      - added Sys::Hwloc::hwloc_compare_objects with aliases
        Sys::Hwloc::Topology->compare_objects and
        Sys::Hwloc::Obj->is_same_obj
        to compare the raw C pointer values

      - added constant HWLOC_XSAPI_VERSION
        (returns 0 where HWLOC_API_VERSION returns undef)

      - implemented Cpuset and Bitmap API with some basic helpers
        and handy functions that are not part of the hwloc C API,
        like $set->sprintf_list

      - destructors $topology->destroy, $cpuset->free, $bitmap->free
        set the unlerlying C pointer to NULL after freeing the
        C struct. This helps to circumvent double-free and
        SIGSEGV. We should implement a DESTROY some day. This
        needs the perl objects to be proxies that store the
        C pointer in some hidden property, instead of being
        simple T_PTROBJ.

      - added PODs for Sys::Hwloc::Topology,
        Sys::Hwloc::Obj, Sys::Hwloc::Cpuset, Sys::Hwloc::Bitmap

      - added the "Artistic License 2.0" to LICENSE

0.04  2010/12/14
      - rewrote from scratch using typemapping
      - added OO-ish methods for hwloc_obj_t and hwloc_topology_t objects
      - support for HWLOC_API_VERSION (undef) .. 0x00010100
      - named the module Sys::Hwloc and submitted to CPAN

0.03  2010/12/07
      - additional constants for hwloc_get_obj_data retrieval
          HWLOC_ObjName       string or undef
          HWLOC_ObjMemory     hashref
          HWLOC_ObjChildren   array
      - rewrote hwloc_get_obj_data with direct stack manipulation
      - made 2nd arg to hwloc_get_obj_data optional,
        if not present, return object data as hashref
      - use perl type UV for unsigned C types

0.02  2010/09/11
      - constant definitions created by h2xs 1.23 with options
          -O -n Hwloc /usr/local/hwloc-1.0.2/include/hwloc.h

0.01  2010/08/18
      - started project


Makefile.PL
################################################################################
#
#  Copyright 2010 Zuse Institute Berlin
#
#  This package and its accompanying libraries is free software; you can
#  redistribute it and/or modify it under the terms of the GPL version 2.0,
#  or the Artistic License 2.0. Refer to LICENSE for the full license text.
#
#  Please send comments to kallies@zib.de
#
################################################################################
#
# Makefile for Sys::Hwloc
#
# $Id: Makefile.PL,v 1.23 2010/12/29 18:14:15 bzbkalli Exp $
#
################################################################################

BEGIN { require 5.006; }

use ExtUtils::MakeMaker;
use Config;
use Cwd;

%INFOS   = (
	    NAME               => 'Sys::Hwloc',
	    VERSION_FROM       => 'lib/Sys/Hwloc.pm',
	    AUTHOR             => 'Bernd Kallies <kallies@zib.de>',
	    LICENSE            => 'perl',
	    ABSTRACT           => 'Perl wrapper around the hwloc API',
	    PREREQ_PM          => {
				   'Test::More' => '0.89',
				  },
	    CONFIGURE_REQUIRES => {
				   'ExtUtils::Constant' => 0,
				  },
	   );

# ------------------------------------------------------------------------------
# Configure using pkg-config
# ------------------------------------------------------------------------------

%CONFIGS = (
	    config            => {
				  'prefix'      => '',
				  'exec_prefix' => '',
				  'libdir'      => '',
				  'includedir'  => '',
				  'iflags'      => '',
				  'libs'        => '',
				  'define'      => '',
				 },
	    HWLOC_API_VERSION => undef,
	    HWLOC_HAS_XML     => 0,
	   );

configure();

# ------------------------------------------------------------------------------
# Write makefile
# ------------------------------------------------------------------------------

delete $INFOS{CONFIGURE_REQUIRES} unless eval { ExtUtils::MakeMaker->VERSION(6.52) };
delete $INFOS{LICENSE}            unless eval { ExtUtils::MakeMaker->VERSION(6.31) };

WriteMakefile(
 	      %INFOS,
 	      INC       => "$CONFIGS{config}->{iflags}",
	      DEFINE    => "$CONFIGS{config}->{define}",
              OPTIMIZE  => "$Config{optimize}",
	      LIBS      => "$CONFIGS{config}->{libs}",
	      PM_FILTER => "sed -e s/\@HWLOC_API_VERSION\@/$CONFIGS{HWLOC_API_VERSION}/",
	      depend    => {
			    "Hwloc.c" => "Hwloc.xs hwloc_cpuset.xsh hwloc_bitmap.xsh",
			   },
 	     );

if(eval { require ExtUtils::Constant; 1 }) {
  # If you edit these definitions to change the constants used by this module,
  # you will need to use the generated const-c.inc and const-xs.inc
  # files to replace their "fallback" counterparts before distributing your
  # changes.
  my @names = (
	       { name => 'HWLOC_OBJ_SYSTEM',                  macro => 1 },
	       { name => 'HWLOC_OBJ_MACHINE',                 macro => 1 },
	       { name => 'HWLOC_OBJ_NODE',                    macro => 1 },
	       { name => 'HWLOC_OBJ_SOCKET',                  macro => 1 },
	       { name => 'HWLOC_OBJ_CACHE',                   macro => 1 },
	       { name => 'HWLOC_OBJ_CORE',                    macro => 1 },
	       { name => 'HWLOC_OBJ_MISC',                    macro => 1 },
	       { name => 'HWLOC_TOPOLOGY_FLAG_IS_THISSYSTEM', macro => 1 },
	       { name => 'HWLOC_TOPOLOGY_FLAG_WHOLE_SYSTEM',  macro => 1 },
	       { name => 'HWLOC_TYPE_DEPTH_MULTIPLE',         macro => 1 },
	       { name => 'HWLOC_TYPE_DEPTH_UNKNOWN',          macro => 1 },
	       { name => 'HWLOC_TYPE_UNORDERED',              macro => 1 },
	       { name => 'HWLOC_HAS_XML',                     macro => 1 },
	       { name => 'HWLOC_CPUBIND_PROCESS',             macro => 1 },
	       { name => 'HWLOC_CPUBIND_THREAD',              macro => 1 },
	       { name => 'HWLOC_CPUBIND_STRICT',              macro => 1 },
	      );

  if(! $CONFIGS{HWLOC_API_VERSION}) {
    push @names, { name => 'HWLOC_OBJ_PROC',           macro => 1 };
  } else {
    push @names, { name => 'HWLOC_OBJ_GROUP',          macro => 1 };
    push @names, { name => 'HWLOC_OBJ_PU',             macro => 1 };
    if($CONFIGS{HWLOC_API_VERSION} ge '0x00010100') {
      push @names, { name => 'HWLOC_CPUBIND_NOMEMBIND',   macro => 1 };
      push @names, { name => 'HWLOC_MEMBIND_PROCESS',     macro => 1 };
      push @names, { name => 'HWLOC_MEMBIND_THREAD',      macro => 1 };
      push @names, { name => 'HWLOC_MEMBIND_STRICT',      macro => 1 };
      push @names, { name => 'HWLOC_MEMBIND_MIGRATE',     macro => 1 };
      push @names, { name => 'HWLOC_MEMBIND_NOCPUBIND',   macro => 1 };
      push @names, { name => 'HWLOC_MEMBIND_DEFAULT',     macro => 1 };
      push @names, { name => 'HWLOC_MEMBIND_FIRSTTOUCH',  macro => 1 };
      push @names, { name => 'HWLOC_MEMBIND_BIND',        macro => 1 };
      push @names, { name => 'HWLOC_MEMBIND_INTERLEAVE',  macro => 1 };
      push @names, { name => 'HWLOC_MEMBIND_REPLICATE',   macro => 1 };
      push @names, { name => 'HWLOC_MEMBIND_NEXTTOUCH',   macro => 1 };
    }
  }

  ExtUtils::Constant::WriteConstants(
                                     NAME         => 'Sys::Hwloc',
                                     NAMES        => \@names,
                                     DEFAULT_TYPE => 'IV',
                                     C_FILE       => 'const-c.inc',
                                     XS_FILE      => 'const-xs.inc',
                                  );

} else {
  use File::Copy;
  use File::Spec;
  foreach my $file ('const-c.inc', 'const-xs.inc') {
    my $fallback = File::Spec->catfile('fallback', $file);
    copy ($fallback, $file) or die "Can't copy $fallback to $file: $!";
  }
}

# ------------------------------------------------------------------------------
# Assemble %CONFIGS
# ------------------------------------------------------------------------------

sub configure {

  print "checking for hwloc... ";
  my $output = `pkg-config --exists --print-errors hwloc 2>&1`;
  if($?) {
    die "no\n$output";
  } else {
    print "yes\n";
  }

  print "checking for hwloc prefix... ";
  $CONFIGS{config}->{prefix} = pkg_config_get("variable=prefix");
  print "$CONFIGS{config}->{prefix}\n";

  print "checking for hwloc eprefix... ";
  $CONFIGS{config}->{eprefix} = pkg_config_get("variable=exec_prefix");
  print "$CONFIGS{config}->{eprefix}\n";

  print "checking for hwloc includedir... ";
  $CONFIGS{config}->{includedir} = pkg_config_get("variable=includedir");
  $CONFIGS{config}->{includedir} = "$CONFIGS{config}->{eprefix}/include" if(! $CONFIGS{config}->{includedir} && $CONFIGS{config}->{eprefix});
  $CONFIGS{config}->{includedir} = "$CONFIGS{config}->{prefix}/include"  if(! $CONFIGS{config}->{includedir} && $CONFIGS{config}->{prefix});
  print "$CONFIGS{config}->{includedir}\n";

  print "checking for hwloc libdir... ";
  $CONFIGS{config}->{libdir}  = pkg_config_get("variable=libdir");
  $CONFIGS{config}->{libdir}  = "$CONFIGS{config}->{eprefix}/lib" if(! $CONFIGS{config}->{libdir} && $CONFIGS{config}->{eprefix});
  $CONFIGS{config}->{libdir}  = "$CONFIGS{config}->{prefix}/lib"  if(! $CONFIGS{config}->{libdir} && $CONFIGS{config}->{prefix});
  print "$CONFIGS{config}->{libdir}\n";

  print "checking for hwloc -I flags... ";
  $CONFIGS{config}->{iflags} = pkg_config_get("cflags-only-I");
  print "$CONFIGS{config}->{iflags}\n";

  print "checking for hwloc libs... ";
  $CONFIGS{config}->{libs} = pkg_config_get("libs");
  print "$CONFIGS{config}->{libs}\n";

  print "checking if hwloc supports hwloc_get_api_version... ";
  if(hwloc_has_func("hwloc_get_api_version")) {
    print "yes\n";
    $CONFIGS{config}->{define} .= " -DHAVE_HWLOC_GET_API_VERSION";
  } else {
    print "no\n";
  }

  print "checking if hwloc supports XML... ";
  if(hwloc_has_func("hwloc_topology_export_xml")) {
    print "yes\n";
    $CONFIGS{HWLOC_HAS_XML} = 1;
  } else {
    print "no\n";
  }
  $CONFIGS{config}->{define} .= " -DHWLOC_HAS_XML=$CONFIGS{HWLOC_HAS_XML}";

  print "checking HWLOC_API_VERSION... ";
  my $cmd  = "$Config{cpprun} -include hwloc.h";
  $cmd    .= " -I$CONFIGS{config}->{includedir}" if $CONFIGS{config}->{includedir};
  $output  = `echo HWLOC_API_VERSION | $cmd $Config{cpplast}`;
  if($?) {
    die "Failed to determine HWLOC_API_VERSION\n";
  } else {
    (my $lastword) = ($output =~ /(\S+)$/);
    die "\nFailed to determine HWLOC_API_VERSION\n" unless defined $lastword;
    if($lastword ne "HWLOC_API_VERSION") {
      $CONFIGS{HWLOC_API_VERSION} = $lastword;
      print "$CONFIGS{HWLOC_API_VERSION}\n";
    } else {
      print "(undef)\n";
    }
  }

}

# ------------------------------------------------------------------------------
# Run pkg-config for hwloc package, retrieve some info
# ------------------------------------------------------------------------------

sub pkg_config_get {
  my $what = shift;
  my $cmd  = "pkg-config --$what hwloc";
  my $data = `$cmd`;
  die "$cmd failed.\n" if $?;
  $data =~ s/[\015\012]+$//;
  return $data;
}

# ------------------------------------------------------------------------------
# See if a function is present in libhwloc.
# Needs $Config{cc} to work.
# Needs $CONFIGS{config} to be set.
# Return 1 on success, 0 on error.
# ------------------------------------------------------------------------------

sub hwloc_has_func {
  my $func = shift;
  return 0 unless $func;

  my $startDir = cwd();
  my $testDir  = '.testlink';
  my $testFile = 't.c';

  rm_fr($testDir);
  mkdir($testDir, 0777) or die "Failed to create dir $testDir: $!";
  chdir($testDir);

  open(FILE, ">$testFile") or die "Failed to write to file $testFile: $!";
  print FILE <<EOT;
char $func ();
int main () { $func (); return 0; }
EOT
  close(FILE);
  `$Config{cc} $CONFIGS{config}->{libs} $testFile >/dev/null 2>&1`;
  my $rc = $?;

  chdir($startDir);
  rm_fr($testDir);

  return $rc ? 0 : 1;
}

# ------------------------------------------------------------------------------
# rm -fr
# ------------------------------------------------------------------------------

sub rm_fr {
  my @files     = @_;
  my @realFiles = ();
  foreach(@files) {
    push @realFiles, glob($_);
  }
  foreach(@realFiles) {
    if(-d $_) {
      rm_fr("$_/*");
      rmdir($_) or die "Failed to remove $_: $!";
    } else {
      chmod(0777, $_);
      unlink($_);
    }
  }
}


